airbyte-workers/src/main/java/io/airbyte/workers/protocols/airbyte/StateDeltaTracker.java

```diff
@@ -0,0 +1,212 @@
+ * both committed and total record counts per stream.
+ * <p>
+ * The StateDeltaTracker is initialized with a memory limit. If this memory limit is exceeded, new
+ * states deltas will not be added and per-stream record counts will not be able to be computed.
+ * This is to prevent OutOfMemoryErrors from crashing the sync.
+ */
+@Slf4j
+public class StateDeltaTracker {
+
+  private static final int STATE_HASH_BYTES = Integer.BYTES;
+  private static final int STREAM_INDEX_BYTES = Short.BYTES;
+  private static final int RECORD_COUNT_BYTES = Long.BYTES;
+  private static final int BYTES_PER_STREAM = STREAM_INDEX_BYTES + RECORD_COUNT_BYTES;
+
+  private final ByteBuffer stateHashByteBuffer;
+  private final ByteBuffer streamIndexByteBuffer;
+  private final ByteBuffer recordCountByteBuffer;
+  private final Set<Integer> committedStateHashes;
+  private final HashMap<Short, Long> committedRecordsByStream;
+
+  /**
+   * Every time a state is added, a new byte[] containing the state hash and per-stream delta will be
+   * added to this list. Every time a state is committed, state deltas up to the committed state are
+   * removed from the head of the list and aggregated into the committed count map. The source thread
+   * adds while the destination thread removes, so those respective methods are synchronized to
+   * prevent thread-safety issues.
+   */
+  @VisibleForTesting
+  protected final List<byte[]> stateDeltas;
+
+  @VisibleForTesting
+  protected long remainingCapacity;
+  @VisibleForTesting
+  protected boolean capacityExceeded;
+
+  public StateDeltaTracker(final long memoryLimitBytes) {
+    this.stateHashByteBuffer = ByteBuffer.allocate(STATE_HASH_BYTES);
+    this.streamIndexByteBuffer = ByteBuffer.allocate(STREAM_INDEX_BYTES);
+    this.recordCountByteBuffer = ByteBuffer.allocate(RECORD_COUNT_BYTES);
+    this.committedStateHashes = new HashSet<>();
+    this.committedRecordsByStream = new HashMap<>();
+    this.stateDeltas = new ArrayList<>();
+    this.remainingCapacity = memoryLimitBytes;
+    this.capacityExceeded = false;
+  }
+
+  /**
+   * Converts the given state hash and per-stream record count map into a {@code byte[]} and stores
+   * it.
+   *
+   * This method is synchronized to provide thread safety between the source thread calling addState
+   * while the destination thread calls commitStateHash.
+   *
+   * @throws CapacityExceededException thrown when the memory footprint of stateDeltas exceeds
+   *         available capacity.
+   */
+  public synchronized void addState(final int stateHash, final Map<Short, Long> streamIndexToRecordCount) throws CapacityExceededException {
+    final int size = STATE_HASH_BYTES + (streamIndexToRecordCount.size() * BYTES_PER_STREAM);
+
+    if (capacityExceeded || remainingCapacity < size) {
+      capacityExceeded = true;
+      throw new CapacityExceededException("Memory capacity is exceeded for StateDeltaTracker.");
+    }
+
+    final byte[] delta = new byte[size];
```

Comment:
see this suggestion PR: https://github.com/airbytehq/airbyte/pull/9331 for how you might be able to simplify the logic for interacting with byte arrays here.